<a href="https://colab.research.google.com/github/GR0409/Statistical-Learning-e21091/blob/main/Assignment_4_Numerical_Data_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Foundations of Statistical Inference & Hypothesis Testing**

## **Part A: Theoretical Fundamentals (Maximum Likelihood & Decision Space)**



1. Consider the sequence of i.i.d. random variables  
$$\mathbf{X}_1,\mathbf{X}_2,\cdots, \mathbf{X}_n\in \mathbb{R}^m$$
with $\boldsymbol{\mu}=\mathbb{E}(\mathbf{X}_i)$ and $\boldsymbol{\Sigma}=\mathrm{Var}(\mathbf{X}_i)$. The unobservable multivariate population parameters $(\boldsymbol{\mu}, \boldsymbol{\Sigma})$ are estimated with their corresponding unbiased empirical sample estimators $(\widehat{\boldsymbol{\mu}}_n, \mathbf{S})$, where
$$\widehat{\boldsymbol{\mu}}_n = \frac{1}{n}\sum_{i=1}^n \mathbf{X}_{i},$$
and
$$\mathbf{S} = \frac{1}{n-1} \sum_{i=1}^n (\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)(\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)^T .$$

* Formally derive why the raw Maximum Likelihood Estimator for the covariance matrix, $\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}$, is inherently **biased** in finite samples, by showing that:

$$\mathbb{E}[\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}] = \frac{n-1}{n}\boldsymbol{\Sigma}$$


* Mathematically demonstrate how applying Bessel's correction yields the unique, unbiased estimator $\mathbf{S}$.


2. Under the null hypothesis $H_0$ that a structural asset remains completely healthy (the status quo), a chosen test statistic tracks a deterministic parametric distribution.
* Define a **Type I Error ($\alpha$)** and a **Type II Error ($\beta$)** explicitly within the context of continuous structural health monitoring.
* If an engineer sets an ultra-conservative significance threshold (e.g., $\alpha = 0.0001$) to prevent false alarms from shutting down an industrial plant, explain the mathematical consequence this decision exerts on the statistical power of the diagnostic layer. How does this modify the geometric volume of your "healthy operation" confidence ellipsoid?

<br>

**Answers**

1.
  
* The uncorrected MLE for the covariance matrix calculates the squared deviations relative to the empirical sample mean $\widehat{\boldsymbol{\mu}}_n$, not the true population mean $\boldsymbol{\mu}$.

Because the sample mean is intrinsically positioned in the exact center of the specific sample data, it minimizes the sum of squared deviations.

Thus, calculating variance around $\widehat{\boldsymbol{\mu}}_n$ artificially shrinks the variance compared to calculating it around the true, unknown $\boldsymbol{\mu}$.

So,

$$\mathbb{E} \left[ \sum_{i=1}^n (\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)(\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)^T \right] = (n - 1)\boldsymbol{\Sigma}$$

Divides by $n$,

$$\mathbb{E}[\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}] = \frac{1}{n} (n - 1)\boldsymbol{\Sigma} = \frac{n-1}{n}\boldsymbol{\Sigma}$$

<br>

* Bessel's Correction

Multiplying the MLE estimator by $\frac{n}{n-1}$,

$$\mathbf{S} = \frac{n}{n-1} \widehat{\boldsymbol{\Sigma}}_{\text{MLE}} = \frac{n}{n-1} \left( \frac{1}{n} \sum (\dots) \right) = \frac{1}{n-1} \sum_{i=1}^n (\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)(\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)^T$$

<br>

2.

* Type I/II Errors

Type I Error ($\alpha$) - A false alarm. Concluding that the structural asset is damaged when it is actually perfectly healthy.

Type II Error ($\beta$) - A missed detection. Concluding that the asset is healthy (status quo) when it has actually suffered an internal fracture or degradation.

<br>

* Statistical significance ($\alpha$) and statistical power ($1 - \beta$) operate on a seesaw. By setting an ultra conservative threshold of $\alpha = 0.0001$, the engineer effectively eliminates false alarms. However, this mathematically destroys the model's statistical power. Geometrically, this inflates the $m$ dimensional operation confidence ellipsoid to a massive volume. Real structural anomalies and subtle drifts will easily fit inside this bloated boundary, resulting in a dangerously high rate of Type II errors.


## **Part B: Theoretical Extension (Asymptotic Distributions & Slutsky's Theorem)**

The Lindeberg-Lévy Multivariate Central Limit Theorem establishes that as the sample snapshot size approaches infinity ($n \to \infty$), the standardized difference between the empirical mean and population mean converges in distribution to a Multivariate Gaussian:

$$\sqrt{n}(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu}) \xrightarrow{d} \mathscr{N}(\mathbf{0}, \boldsymbol{\Sigma})$$



However, the true population covariance matrix $\boldsymbol{\Sigma}$ is a latent, unobservable parameter.
* State **Slutsky’s Theorem** rigorously.
* Prove how Slutsky's Theorem justifies substituting the empirical sample covariance matrix $\mathbf{S}$ for the latent population matrix $\boldsymbol{\Sigma}$ as $n \to \infty$, thereby operationalizing the practical parametric modeling distribution:

$$\widehat{\boldsymbol{\mu}}_n \sim \mathscr{N}\left({\boldsymbol{\mu}}, \frac{1}{n}\mathbf{S}\right)$$







<br>

**Answers**

* Slutsky's Theorem,

 If a sequence of random vectors $\mathbf{X}_n$ converges in distribution to $\mathbf{X}$ , and a sequence of random matrices $\mathbf{Y}_n$ converges in probability to a constant matrix $\mathbf{c}$, then the product converges in distribution is $\mathbf{Y}_n \mathbf{X}_n \xrightarrow{d} \mathbf{c}\mathbf{X}$.

<br>

*  The Central Limit Theorem gives that,

$\sqrt{n}(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu}) \xrightarrow{d} \mathscr{N}(\mathbf{0}, \boldsymbol{\Sigma})$.

$\boldsymbol{\Sigma}$ is unobservable.

By the Law of Large Numbers,

$\mathbf{S} \xrightarrow{p} \boldsymbol{\Sigma}$ as $n \to \infty$.

Using the Continuous Mapping Theorem,

$\mathbf{S}^{-1/2} \xrightarrow{p} \boldsymbol{\Sigma}^{-1/2}$.

Applying Slutsky’s theorem,

$$\mathbf{S}^{-1/2} \sqrt{n}(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu}) \xrightarrow{d} \mathscr{N}(\mathbf{0}, \mathbf{I})$$

By rearranging,

$$\widehat{\boldsymbol{\mu}}_n \sim \mathscr{N}\left({\boldsymbol{\mu}}, \frac{1}{n}\mathbf{S}\right)$$

<br>



## **Part C: Numerical Verification**

***Scenario:** You are provided an industrial dataset containing $n = 5,000$ snapshots across $m = 3$ sensor channels measuring physical strain. To verify the weak i.i.d. assumption before initializing a global PCA or Factor Analysis layer, you must explicitly test **Condition 1: Constant Joint Mean (First Moment Homogeneity)** across the time index.*

**Your Task:**
Write a clean, self-contained Python function within Google Colab to partition the dataset into $g = 5$ consecutive, non-overlapping blocks (chunks) of size $n_j = 1,000$. Implement a Multivariate Analysis of Variance (MANOVA) tracking framework from scratch (or leveraging `statsmodels`) to calculate **Wilks' Lambda ($\Lambda$)**.

$$\Lambda = \frac{|\mathbf{W}|}{|\mathbf{B} + \mathbf{W}|}$$

Transform your calculated $\Lambda$ using **Bartlett’s $\chi^2$ approximation**:


$$\chi^2_{\text{calc}} = -\left(n - 1 - \frac{m + g}{2}\right) \ln \Lambda$$

Execute your code using the synthetic data generator provided below. Report your exact calculated Wilks' $\Lambda$, the Bartlett $\chi^2$ test statistic, and its corresponding $p$-value under $\chi^2 \sim \chi^2_{m(g-1)}$. Based on an alpha threshold of $\alpha = 0.05$, state your final diagnostic conclusion: Has the baseline center shifted over time, or is the first moment homogeneous?



```python
import numpy as np
import pandas as pd
import scipy.stats as stats

# --- DO NOT MODIFY: Synthetic Strain Sensor Data Generator ---
np.random.seed(42)
n_samples = 5000
n_features = 3

# Generating base stationary multivariate normal noise
base_data = np.random.multivariate_normal(
    mean=[0.5, -0.2, 1.1],
    cov=[[0.09, 0.02, 0.01],
         [0.02, 0.06, 0.03],
         [0.01, 0.03, 0.05]],
    size=n_samples
)

# Injecting a subtle, hidden structural drift in the final 1,000 snapshots
base_data[4000:, 0] += 0.015  # Sensor 1 drift
base_data[4000:, 2] -= 0.010  # Sensor 3 drift

df_strain = pd.DataFrame(base_data, columns=['Strain_Ch1', 'Strain_Ch2', 'Strain_Ch3'])
# ----------------------------------------------------------------

def verify_first_moment_homogeneity(df: pd.DataFrame, g_chunks: int = 5) -> dict:
    n, m = df.shape
    chunk_size = n // g_chunks
    
    # Global Mean
    global_mean = df.mean().values
    
    # Initialize Within (W) and Between (B) matrices
    W = np.zeros((m, m))
    B = np.zeros((m, m))
    
    # Iterate through chunks
    for i in range(g_chunks):
        chunk = df.iloc[i*chunk_size : (i+1)*chunk_size].values
        chunk_mean = chunk.mean(axis=0)
        
        # Calculate W: Sum of squared deviations within this specific chunk
        chunk_centered = chunk - chunk_mean
        W += chunk_centered.T @ chunk_centered
        
        # Calculate B: Deviation of chunk mean from global mean, scaled by chunk size
        mean_diff = (chunk_mean - global_mean).reshape(-1, 1)
        B += chunk_size * (mean_diff @ mean_diff.T)
        
    # Calculate Wilks' Lambda
    wilks_lambda = np.linalg.det(W) / np.linalg.det(B + W)
    
    # Bartlett's Chi-Square Approximation
    chi2_stat = -(n - 1 - (m + g_chunks) / 2) * np.log(wilks_lambda)
    
    # Degrees of Freedom and p-value
    df_chi2 = m * (g_chunks - 1)
    p_value = 1 - stats.chi2.cdf(chi2_stat, df_chi2)
    
    return {
        "Wilks_Lambda": wilks_lambda,
        "Chi2_Statistic": chi2_stat,
        "p_value": p_value,
        "Diagnostic": "Homogeneous" if p_value > 0.05 else "Drift Detected"
    }

In [7]:
import numpy as np
import pandas as pd
import scipy.stats as stats

# --- DO NOT MODIFY: Synthetic Strain Sensor Data Generator ---
np.random.seed(42)
n_samples = 5000
n_features = 3

# Generating base stationary multivariate normal noise
base_data = np.random.multivariate_normal(
    mean=[0.5, -0.2, 1.1],
    cov=[[0.09, 0.02, 0.01],
         [0.02, 0.06, 0.03],
         [0.01, 0.03, 0.05]],
    size=n_samples
)

# Injecting a subtle, hidden structural drift in the final 1,000 snapshots
base_data[4000:, 0] += 0.015  # Sensor 1 drift
base_data[4000:, 2] -= 0.010  # Sensor 3 drift

df_strain = pd.DataFrame(base_data, columns=['Strain_Ch1', 'Strain_Ch2', 'Strain_Ch3'])
# ----------------------------------------------------------------

def verify_first_moment_homogeneity(df: pd.DataFrame, g_chunks: int = 5) -> dict:
    n, m = df.shape
    chunk_size = n // g_chunks

    # Global Mean
    global_mean = df.mean().values

    # Initialize Within (W) and Between (B) matrices
    W = np.zeros((m, m))
    B = np.zeros((m, m))

    # Iterate through chunks
    for i in range(g_chunks):
        chunk = df.iloc[i*chunk_size : (i+1)*chunk_size].values
        chunk_mean = chunk.mean(axis=0)

        # Calculate W: Sum of squared deviations within this specific chunk
        chunk_centered = chunk - chunk_mean
        W += chunk_centered.T @ chunk_centered

        # Calculate B: Deviation of chunk mean from global mean, scaled by chunk size
        mean_diff = (chunk_mean - global_mean).reshape(-1, 1)
        B += chunk_size * (mean_diff @ mean_diff.T)

    # Calculate Wilks' Lambda
    wilks_lambda = np.linalg.det(W) / np.linalg.det(B + W)

    # Bartlett's Chi-Square Approximation
    chi2_stat = -(n - 1 - (m + g_chunks) / 2) * np.log(wilks_lambda)

    # Degrees of Freedom and p-value
    df_chi2 = m * (g_chunks - 1)
    p_value = 1 - stats.chi2.cdf(chi2_stat, df_chi2)

    return {
        "Wilks_Lambda": wilks_lambda,
        "Chi2_Statistic": chi2_stat,
        "p_value": p_value,
        "Diagnostic": "Homogeneous" if p_value > 0.05 else "Drift Detected"
    }

# --- THIS IS THE MISSING EXECUTION BLOCK ---
# Call the function and print the results
results = verify_first_moment_homogeneity(df_strain, g_chunks=5)

print("\n--- Diagnostic Results ---")
for key, value in results.items():
    if isinstance(value, float):
        print(f"{key}: {value:.6e}" if key == 'p_value' else f"{key}: {value:.4f}")
    else:
        print(f"{key}: {value}")


--- Diagnostic Results ---
Wilks_Lambda: 0.9905
Chi2_Statistic: 47.9215
p_value: 3.225526e-06
Diagnostic: Drift Detected


# **Geometric Subspace Optimization via Principal Component Analysis (PCA)**

## **Part A: Theoretical Fundamentals (Coordinate Projections & Orthogonality)**

1. Let $\widetilde{\mathbf{X}}_i = (\mathbf{X}_i - \boldsymbol{\mu}) \in \mathbb{R}^m$ represent a zero-mean, centered baseline feature vector with a symmetric, positive-semidefinite population covariance matrix $\boldsymbol{\Sigma} = \mathbb{E}[\widetilde{\mathbf{X}}_i \widetilde{\mathbf{X}}_i^T]$. We apply the spectral decomposition $\boldsymbol{\Sigma} = \mathbf{P} \mathbf{\Lambda} \mathbf{P}^T$, and define the orthogonal projection mapping:

$$\mathbf{Z}_i = \mathbf{P}^T \widetilde{\mathbf{X}}_i$$


* Prove step-by-step that the theoretical covariance matrix of the transformed vector, $\mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T]$, simplifies exactly to the strictly diagonal matrix $\mathbf{\Lambda}$.
* Explain the precise geometric and statistical meaning of this diagonal result regarding the cross-covariances between distinct coordinates $Z_{i,j}$ and $Z_{i,k}$ (for $j \neq k$).


2. The total variance of a multivariate system can be measured in its raw coordinate space or its transformed principal component space.
* Utilizing the algebraic properties of the matrix trace ($\text{tr}$), specifically its invariance under cyclic permutations, formally prove that the total system variance is preserved during the mapping from raw space to the de-correlated principal space:

$$\text{tr}(\boldsymbol{\Sigma}) = \text{tr}(\mathbf{\Lambda}) = \sum_{j=1}^m \lambda_j$$


* If we partition the empirical matrix into an explained subspace ($\widehat{\mathbf{P}}_k \in \mathbb{R}^{m \times k}$) and an unexplained residual subspace ($\widehat{\mathbf{P}}_{m-k} \in \mathbb{R}^{m \times (m-k)}$), state the rigorous mathematical formulations for the **Cumulative Explained Variance Ratio $\Phi(k)$** and the **Residual Unexplained Variance Ratio $\Psi(k)$**.


3. In structural health monitoring (SHM), we truncate our coordinate system to retain only the first $k$ components ($k < m$). The centered raw vector is partitioned as $(\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n) = \widehat{\mathbf{P}}_k \mathbf{z}_{i,k} + \widehat{\mathbf{P}}_{m-k} \mathbf{z}_{i,m-k}$.
* Define the reconstructed vector $\widehat{\mathbf{x}}_i$ and the residual error vector $\mathbf{e}_i$.
* Prove via the Pythagorean identity that the total squared Euclidean norm of the centered observation vector cleanly decomposes into independent subspace energies:

$$\|\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n\|^2 = \|\mathbf{z}_{i,k}\|^2 + \|\mathbf{z}_{i,m-k}\|^2$$


* Contrast the diagnostic functions of Hotelling's $T^2$ statistic and the residual $Q$ statistic (SPE). If a structure encounters an environmental load anomaly (e.g., extreme operational wind loads) versus an internal structural fracture (e.g., a fatigue crack changing the structural load pathways), match each event type to the statistic ($T^2$ or $Q$) most sensitive to it and justify your choices.

<br>

**Answers**

1.

* Proof of Diagonal Simplification,

$$\mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T] = \mathbb{E}[(\mathbf{P}^T \widetilde{\mathbf{X}}_i) (\mathbf{P}^T \widetilde{\mathbf{X}}_i)^T]$$

$$= \mathbb{E}[\mathbf{P}^T \widetilde{\mathbf{X}}_i \widetilde{\mathbf{X}}_i^T \mathbf{P}]$$

$$= \mathbf{P}^T \mathbb{E}[\widetilde{\mathbf{X}}_i \widetilde{\mathbf{X}}_i^T] \mathbf{P}$$

Since $\mathbb{E}[\widetilde{\mathbf{X}}_i \widetilde{\mathbf{X}}_i^T] = \boldsymbol{\Sigma}$,

we substitute $\boldsymbol{\Sigma} = \mathbf{P} \mathbf{\Lambda} \mathbf{P}^T$,

$$= \mathbf{P}^T (\mathbf{P} \mathbf{\Lambda} \mathbf{P}^T) \mathbf{P}$$Because eigenvectors are orthogonal ($\mathbf{P}^T \mathbf{P} = \mathbf{I}$),

$$= \mathbf{I} \mathbf{\Lambda} \mathbf{I} = \mathbf{\Lambda}$$

<br>

* Because $\mathbf{\Lambda}$ is strictly a diagonal matrix, all off-diagonal elements are exactly zero. Statistically, this guarantees that the extracted principal components are entirely independent. Geometrically, it means the new coordinate axes are perpendicular to each other.

<br>

2.

* $$\text{tr}(\boldsymbol{\Sigma}) = \text{tr}(\mathbf{P} \mathbf{\Lambda} \mathbf{P}^T)$$

Using the cyclic permutation property of the trace ($\text{tr}(ABC) = \text{tr}(CAB)$),

$$\text{tr}(\mathbf{P} \mathbf{\Lambda} \mathbf{P}^T) = \text{tr}(\mathbf{P}^T \mathbf{P} \mathbf{\Lambda})$$

Since $\mathbf{P}^T \mathbf{P} = \mathbf{I}$:$$\text{tr}(\mathbf{\Lambda}) = \sum_{j=1}^m \lambda_j$$

<br>

* umulative Explained $\Phi(k)$,

$$\frac{\sum_{j=1}^k \lambda_j}{\sum_{j=1}^m \lambda_j}$$

Residual Unexplained $\Psi(k)$,

$$\frac{\sum_{j=k+1}^m \lambda_j}{\sum_{j=1}^m \lambda_j} = 1 - \Phi(k)$$

<br>

3.

* $\widehat{\mathbf{x}}_i$ is the projection of the data strictly within the healthy expected bounds.

  The error $\mathbf{e}_i$ represents the unpredictable noise discarded by the model.

<br>

* Because the principal subspace $\widehat{\mathbf{P}}_k$ and the residual subspace $\widehat{\mathbf{P}}_{m-k}$ are strictly orthogonal, their cross-product is zero.

So total energy,

$$\|\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n\|^2 = \|\mathbf{z}_{i,k}\|^2 + \|\mathbf{z}_{i,m-k}\|^2$$

<br>

* Diagnostic Matching,

Extreme Wind Load - Hotelling's $T^2$,

Wind load operates along the known physical axes of the structure. It doesn't change how the structure bends, it just bends it more. $T^2$ tracks amplitude variations within the expected principal subspace.

Internal Fracture - $Q$ Statistic,

A fracture breaks the structural stiffness matrix, creating a totally new physical pathway that violates the baseline correlation models. This causes the signal to jump out of the expected geometric plane, causing a massive spike in the residual $Q$ (SPE) statistic.



## **Part B: Numerical Verification**



***Scenario:** You are building a real-time diagnostics layer for a structural component monitored by $m = 4$ sensors. You have a baseline archive of $n = 3,000$ nominal snapshots. To establish operational boundaries, you must write an optimization pipeline that analyzes how Hotelling’s $T^2$ and the residual $Q$ statistic respond to different truncation boundaries $k \in \{1, 2, 3\}$.*

**Your Task:**
Complete the Python class method below inside Google Colab. The function must:

1. Standardize and center the incoming realizations.
2. Decompose the unbiased sample covariance matrix $\mathbf{S}$ using `np.linalg.eigh` and sort components in descending order ($\widehat{\lambda}_1 \ge \dots \ge \widehat{\lambda}_m$).
3. Loop through every possible subspace cutoff value $k \in \{1, 2, 3\}$, calculating the sample vector array for Hotelling’s $T^2$ ($T^2_i = \sum_{j=1}^k \frac{z_{i,j}^2}{\widehat{\lambda}_j}$) and the $Q$ statistic ($Q_i = \sum_{j=k+1}^m z_{i,j}^2$).
4. Compute the expected value (empirical mean) of $T^2$ and $Q$ across all samples for each $k$.
5. Return these metrics to verify your theoretical derivations in Part A.



<br>

**Answer**


In [8]:
import numpy as np
import pandas as pd

class SubspaceDiagnostics:
    def __init__(self, df: pd.DataFrame):
        """Initializes the diagnostics pipeline with an incoming DataFrame."""
        self.raw_data = df.values
        self.n, self.m = self.raw_data.shape

    def optimize_subspace_boundaries(self) -> dict:
        """
        Calculates the expected values of Hotelling's T^2 and the Q statistic
        across truncation boundaries k in {1, 2, 3}.
        """
        # Step 1: Standardize and center the incoming realizations (ddof=1 for unbiased)
        mu = np.mean(self.raw_data, axis=0)
        sigma = np.std(self.raw_data, axis=0, ddof=1)
        Z = (self.raw_data - mu) / sigma

        # Step 2: Decompose the unbiased sample covariance matrix S
        S = np.cov(Z, rowvar=False)
        eigenvalues, eigenvectors = np.linalg.eigh(S)

        # np.linalg.eigh returns ascending order; we must sort descending
        idx = np.argsort(eigenvalues)[::-1]
        eigenvalues = eigenvalues[idx]
        eigenvectors = eigenvectors[:, idx]

        # Map standardized data into the full principal component space (Z_pca = Z * P)
        Z_pca = Z @ eigenvectors

        results = {}

        # Step 3 & 4: Loop through every possible subspace cutoff value k
        for k in [1, 2, 3]:
            # Calculate Hotelling's T^2 array for the retained k components
            # T^2_i = sum_{j=1}^k (z_{i,j}^2 / lambda_j)
            T2_array = np.sum((Z_pca[:, :k]**2) / eigenvalues[:k], axis=1)

            # Calculate the Q (SPE) statistic array for the residual (m-k) components
            # Q_i = sum_{j=k+1}^m (z_{i,j}^2)
            Q_array = np.sum(Z_pca[:, k:]**2, axis=1)

            # Compute the expected value (empirical mean) across all samples
            mean_T2 = np.mean(T2_array)
            mean_Q = np.mean(Q_array)

            # Step 5: Return metrics
            results[f'k={k}'] = {
                'Expected_T2': mean_T2,
                'Expected_Q': mean_Q,
                'Theoretical_Q': np.sum(eigenvalues[k:]) # For verification
            }

        return results

# ==========================================
# Execution & Verification Block
# ==========================================
if __name__ == "__main__":
    # Generate n=3000 nominal snapshots across m=4 sensors
    np.random.seed(42)
    n_samples = 3000

    # Simulating highly correlated sensor data
    f1 = np.random.normal(0, 1, n_samples)
    f2 = np.random.normal(0, 1, n_samples)

    s1 = 0.9 * f1 + np.random.normal(0, 0.1, n_samples)
    s2 = 0.8 * f1 + 0.2 * f2 + np.random.normal(0, 0.1, n_samples)
    s3 = 0.3 * f1 + 0.8 * f2 + np.random.normal(0, 0.1, n_samples)
    s4 = 0.1 * f1 + 0.9 * f2 + np.random.normal(0, 0.2, n_samples)

    df_nominal = pd.DataFrame({'S1': s1, 'S2': s2, 'S3': s3, 'S4': s4})

    # Initialize class and run the optimization method
    diagnostics = SubspaceDiagnostics(df_nominal)
    metrics = diagnostics.optimize_subspace_boundaries()

    # Print results to verify theoretical derivations
    print("--- Subspace Truncation Optimization Results ---")
    for k_val, data in metrics.items():
        print(f"\nBoundary: {k_val}")
        print(f"  -> Empirical Mean T²: {data['Expected_T2']:.4f} (Theoretical T² = {k_val[-1]})")
        print(f"  -> Empirical Mean Q : {data['Expected_Q']:.4f} (Theoretical Q  = {data['Theoretical_Q']:.4f})")

--- Subspace Truncation Optimization Results ---

Boundary: k=1
  -> Empirical Mean T²: 0.9997 (Theoretical T² = 1)
  -> Empirical Mean Q : 1.3775 (Theoretical Q  = 1.3780)

Boundary: k=2
  -> Empirical Mean T²: 1.9993 (Theoretical T² = 2)
  -> Empirical Mean Q : 0.0423 (Theoretical Q  = 0.0423)

Boundary: k=3
  -> Empirical Mean T²: 2.9990 (Theoretical T² = 3)
  -> Empirical Mean Q : 0.0140 (Theoretical Q  = 0.0140)


# **Latent Subspace Decomposition via Factor Analysis (FA)**

## **Part A: Theoretical Fundamentals (Generative Model & Commonalities)**

1. Consider your structured data setup where $\mathbf{Z}_i \in \mathbb{R}^m$ is a normalized, zero-mean baseline feature vector whose population covariance matrix represents the universal correlation matrix $\mathbf{R} = \mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T]$. The Factor Analysis model posits the linear generative structure:
$$\mathbf{Z}_i = \boldsymbol{\Lambda} \mathbf{F}_i + \boldsymbol{\epsilon}_i$$
where $\mathbf{F}_i \in \mathbb{R}^k$ are the orthogonal common factors ($\mathbb{E}[\mathbf{F}_i\mathbf{F}_i^T] = \mathbf{I}_{k \times k}$) and $\boldsymbol{\epsilon}_i \in \mathbb{R}^m$ is the diagonalized measurement noise matrix ($\mathbb{E}[\boldsymbol{\epsilon}_i\boldsymbol{\epsilon}_i^T] = \boldsymbol{\Psi} = \text{diag}(\varphi_1^2, \dots, \varphi_m^2)$).
* Utilizing the constraint that the latent factors and errors are independent ($\mathbb{E}[\boldsymbol{\epsilon}_i \mathbf{F}_i^T] = \mathbf{0}$), prove the **Fundamental Equation of Factor Analysis**:

$$\mathbf{R} = \boldsymbol{\Lambda}\boldsymbol{\Lambda}^T + \boldsymbol{\Psi}$$


* Focus on the $j$-th diagonal entry of this decomposition. Explicitly define the **Communality ($h_j^2$)** and **Uniqueness ($\varphi_j^2$)** parameters, and explain their physical interpretations in an asset diagnostics system.


2. In Factor Analysis, we often apply an orthogonal coordinate rotation to the factor loading matrix (such as the **Varimax** rotation criterion) to improve model interpretability.
* Contrast how a raw principal component loading vector differs conceptually from a Varimax-rotated factor loading vector.
* Explain why orthogonal rotations can alter the individual columns of the loading matrix $\boldsymbol{\Lambda}$ without changing the calculated communality ($h_j^2$) of any sensor or modifying the global covariance approximation $\mathbf{R} \approx \boldsymbol{\Lambda}\boldsymbol{\Lambda}^T + \boldsymbol{\Psi}$.

<br>

**Answers**

1.

* Proof,

$$\mathbf{R} = \mathbb{E}[(\boldsymbol{\Lambda} \mathbf{F}_i + \boldsymbol{\epsilon}_i)(\boldsymbol{\Lambda} \mathbf{F}_i + \boldsymbol{\epsilon}_i)^T]$$

Expand the quadratic,

$$= \mathbb{E}[\boldsymbol{\Lambda} \mathbf{F}_i \mathbf{F}_i^T \boldsymbol{\Lambda}^T + \boldsymbol{\Lambda} \mathbf{F}_i \boldsymbol{\epsilon}_i^T + \boldsymbol{\epsilon}_i \mathbf{F}_i^T \boldsymbol{\Lambda}^T + \boldsymbol{\epsilon}_i \boldsymbol{\epsilon}_i^T]$$

Because factors and noise are independent, cross terms vanish ($\mathbb{E}[\boldsymbol{\epsilon}_i \mathbf{F}_i^T] = 0$). Also, $\mathbb{E}[\mathbf{F}_i \mathbf{F}_i^T] = \mathbf{I}$, and $\mathbb{E}[\boldsymbol{\epsilon}_i \boldsymbol{\epsilon}_i^T] = \boldsymbol{\Psi}$.

$$= \boldsymbol{\Lambda} \mathbf{I} \boldsymbol{\Lambda}^T + \boldsymbol{\Psi} = \boldsymbol{\Lambda}\boldsymbol{\Lambda}^T + \boldsymbol{\Psi}$$

<br>

* Communality ($h_j^2$): The portion of variance in a sensor caused by true, underlying physical forces shared across the structure like vibration and load.

  Uniqueness ($\varphi_j^2$): The portion of variance isolated purely to that specific sensor like localized thermal drift and electrical short.

<br>

2.

* PCA packs maximum variance sequentially, causing sensors to load messily across multiple components. Varimax rotates these axes to maximize the variance of the squared loadings. This forces coefficients toward 1 or 0, creating a simple structure where each sensor aligns cleanly with a single specific physical factor.

<br>

* Applying an orthogonal rotation matrix,

$\mathbf{T}$ (where $\mathbf{T}\mathbf{T}^T = \mathbf{I}$) to the loadings yields,

$(\boldsymbol{\Lambda}\mathbf{T})(\boldsymbol{\Lambda}\mathbf{T})^T = \boldsymbol{\Lambda}\mathbf{T}\mathbf{T}^T\boldsymbol{\Lambda}^T = \boldsymbol{\Lambda}\boldsymbol{\Lambda}^T$.

The final mathematical matrix remains identical, we just changed the angle we use to look at it.

<br>




1. Because the common factors $\mathbf{F}_i$ are unobservable latent parameters, they cannot be directly extracted via geometric projections like PCA. Instead, we must treat them as hidden random vectors and calculate their realization using **Thomson’s Regression Method**, which relies on the conditional expectation $\mathbf{f}_i = \mathbb{E}[\mathbf{F}_i | \mathbf{z}_i]$.
* Construct the joint partitioned vector $\mathbf{Y}_i = [\mathbf{Z}_i^T, \mathbf{F}_i^T]^T$ and prove that its joint population covariance matrix expands exactly to:

$$\boldsymbol{\Sigma}_{YY} = \begin{bmatrix} \mathbf{R} & \boldsymbol{\Lambda} \\ \boldsymbol{\Lambda}^T & \mathbf{I}_{k \times k} \end{bmatrix}$$


* Under the standard assumption of a Multivariate Normal distribution, the conditional mean of a partitioned system is governed by the projection identity $\mathbb{E}[\mathbf{X}_2 | \mathbf{x}_1] = \boldsymbol{\mu}_2 + \boldsymbol{\Sigma}_{21} \boldsymbol{\Sigma}_{11}^{-1} (\mathbf{x}_1 - \boldsymbol{\mu}_1)$. Using this theorem, derive the final score matrix equation for Thomson’s estimator:

$$\mathbf{f}_i = \boldsymbol{\Lambda}^T \mathbf{R}^{-1} \mathbf{z}_i$$


<br>

**Answers**

1.

* Deriving $\boldsymbol{\Sigma}_{YY}$,

We know $\mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T] = \mathbf{R}$ and $\mathbb{E}[\mathbf{F}_i \mathbf{F}_i^T] = \mathbf{I}$.

We just need the cross-covariance block,

$$\mathbb{E}[\mathbf{Z}_i \mathbf{F}_i^T] = \mathbb{E}[(\boldsymbol{\Lambda}\mathbf{F}_i + \boldsymbol{\epsilon}_i)\mathbf{F}_i^T] = \boldsymbol{\Lambda} \mathbb{E}[\mathbf{F}_i \mathbf{F}_i^T] + \mathbb{E}[\boldsymbol{\epsilon}_i \mathbf{F}_i^T]$$

Since $\mathbb{E}[\boldsymbol{\epsilon}_i \mathbf{F}_i^T] = 0$ and $\mathbb{E}[\mathbf{F}_i \mathbf{F}_i^T] = \mathbf{I}$, the cross block is simply $\boldsymbol{\Lambda}$.

This yields the block matrix $\boldsymbol{\Sigma}_{YY} = \begin{bmatrix} \mathbf{R} & \boldsymbol{\Lambda} \\ \boldsymbol{\Lambda}^T & \mathbf{I}_{k \times k} \end{bmatrix}$.

<br>

* Using the projection identity,

$\mathbb{E}[\mathbf{X}_2 | \mathbf{x}_1] = \boldsymbol{\mu}_2 + \boldsymbol{\Sigma}_{21} \boldsymbol{\Sigma}_{11}^{-1} (\mathbf{x}_1 - \boldsymbol{\mu}_1)$.

Substitute our terms,

$\mathbf{X}_2$ becomes $\mathbf{F}_i$, $\mathbf{x}_1$ becomes $\mathbf{z}_i$, and both means are $\mathbf{0}$.
$\boldsymbol{\Sigma}_{21}$ is the cross-block $\boldsymbol{\Lambda}^T$, and $\boldsymbol{\Sigma}_{11}^{-1}$ is $\mathbf{R}^{-1}$.

This leaves us exactly with,

$\mathbf{f}_i = \boldsymbol{\Lambda}^T \mathbf{R}^{-1} \mathbf{z}_i$.

<br>


## **Part B: Parametric Factor Partition Engine & UI Architecture**

### Develop a Mathematical Engine & Pre-processing Guardrails


Implement an isolated algorithm to structure and decompose empirical data structures according to the following guidelines:

1. **Z-Score Mapping Layer:** Accept an arbitrary data frame and isolate numeric observation columns. Standardize the observations into standard normal space $Z \in \mathbb{R}^{n \times m}$ applying an unbiased sample standard deviation ($ddof=1$). Build an explicit numerical guardrail that catches zero-variance channels ($\sigma_j = 0$) and caps them to a minimum boundary threshold of $10^{-15}$ to prevent divide-by-zero runtime exceptions.
2. **Dimensionality Constraint Layer:** Let $m$ represent the count of isolated sensor features, and $k$ represent the target count of latent physical factors. Program two validation rule exceptions:
* Reject processing if the operational snapshot count ($n$) is less than or equal to the channel count ($m$).
* Throw an descriptive error if the requested latent space dimension ($k$) is not strictly less than the available feature space dimension ($m$).


3. **Information Decomposition Layer:** Fit a Factor Analysis framework utilizing a Maximum Likelihood or Singular Value Decomposition (SVD) algorithm combined with an orthogonal Varimax rotation matrix ($T$). From the fitted parameters, compute the vector arrays for:
* **Communality Vector ($h^2 \in \mathbb{R}^m$):** The total shared variance captured across each channel.
* **Uniqueness Vector ($\varphi^2 \in \mathbb{R}^m$):** The remaining isolated sensor noise floor.

<br>

**Answer**

In [10]:
pip install factor_analyzer

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.8/42.8 kB 1.7 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  Created wheel for factor_analyzer: filename=factor_analyzer-0.5.1-py2.py3-none-any.whl size=42655 sha256=afb7fe556535b75b0ba63e3f22e024b7f11fa3fffac39544e66081a65ea30658
  Stored in directory: /root/.cache/pip/wheels/a2/af/06/f4d4ed4d9d714fda437fb1583629417319603c2266e7b233cc
Successfully built factor_analyzer


In [13]:
import numpy as np
import pandas as pd
from factor_analyzer import FactorAnalyzer

# 2. The Isolated Algorithm
def decompose_empirical_data(df: pd.DataFrame, k: int) -> dict:

    numeric_df = df.select_dtypes(include=[np.number])
    n, m = numeric_df.shape

    if n <= m:
        raise ValueError(f"Snapshot count (n={n}) must be strictly greater than channel count (m={m}).")

    if k >= m:
        raise ValueError(f"Requested latent space dimension (k={k}) must be strictly less than feature space (m={m}).")

    mu = numeric_df.mean(axis=0)
    sigma = numeric_df.std(axis=0, ddof=1)

    sigma = sigma.replace(0.0, 1e-15)
    Z = (numeric_df - mu) / sigma

    fa = FactorAnalyzer(n_factors=k, rotation='varimax', method='ml')
    fa.fit(Z)

    h2 = fa.get_communalities()
    phi2 = fa.get_uniquenesses()

    return {
        "Z_matrix": Z,
        "communalities": h2,
        "uniquenesses": phi2,
        "loadings_matrix": fa.loadings_
    }

# ==========================================
# 3. Execution & Verification Block
# ==========================================
print("Installing packages and running Factor Analysis...")

# Generate synthetic sensor data to test the function
np.random.seed(42)
n_samples = 3000

# Two hidden physical forces
f1 = np.random.normal(0, 1, n_samples)
f2 = np.random.normal(0, 1, n_samples)

# Four sensors observing the forces
s1 = 0.9 * f1 + np.random.normal(0, 0.1, n_samples)
s2 = 0.8 * f1 + 0.2 * f2 + np.random.normal(0, 0.1, n_samples)
s3 = 0.3 * f1 + 0.8 * f2 + np.random.normal(0, 0.1, n_samples)
# Sensor 4 is mostly just noise (to test the Uniqueness calculation)
s4 = 0.1 * f1 + 0.9 * f2 + np.random.normal(0, 1.5, n_samples)

df_asset = pd.DataFrame({'Sensor_1': s1, 'Sensor_2': s2, 'Sensor_3': s3, 'Sensor_4': s4})

# Run the algorithm (Asking for k=2 latent factors)
results = decompose_empirical_data(df_asset, k=2)

# 4. Print the Output
print("\n--- FA Decomposition Results ---")
print(f"Dataset Shape: {results['Z_matrix'].shape}\n")

print("Communalities (h²) - Shared Structural Variance:")
for i, val in enumerate(results['communalities']):
    print(f"  Sensor_{i+1}: {val:.4f} ({val*100:.1f}%)")

print("\nUniquenesses (φ²) - Isolated Sensor Noise:")
for i, val in enumerate(results['uniquenesses']):
    print(f"  Sensor_{i+1}: {val:.4f} ({val*100:.1f}%)")

print("\nRotated Loadings Matrix (Λ):")
print(np.round(results['loadings_matrix'], 4))

Installing packages and running Factor Analysis...

--- FA Decomposition Results ---
Dataset Shape: (3000, 4)

Communalities (h²) - Shared Structural Variance:
  Sensor_1: 0.9828 (98.3%)
  Sensor_2: 0.9878 (98.8%)
  Sensor_3: 0.9950 (99.5%)
  Sensor_4: 0.2754 (27.5%)

Uniquenesses (φ²) - Isolated Sensor Noise:
  Sensor_1: 0.0172 (1.7%)
  Sensor_2: 0.0122 (1.2%)
  Sensor_3: 0.0050 (0.5%)
  Sensor_4: 0.7246 (72.5%)

Rotated Loadings Matrix (Λ):
[[0.9903 0.0456]
 [0.9529 0.2825]
 [0.2924 0.9537]
 [0.0265 0.5241]]


/usr/local/lib/python3.12/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(


### Latent Score Projection

1. **Mathematical State Reconstruction:** Using Thomson’s Minimum Mean Squared Error (MMSE) regression methodology, transform the standard normal snapshot array ($Z$) directly into hidden factor states $F \in \mathbb{R}^{n \times k}$ using the system relation:

$$F = Z R^{-1} \Lambda_{\text{rotated}}$$


2. **Performance Diagnostic Telemetry:** Calculate the empirical variance of the reconstructed factor scores enforcing an unbiased scale correction ($ddof=1$). Print a clean confirmation breakdown out to the logging console that presents the calculated system-wide metrics:

$$\text{Average System Communality \%} = \frac{1}{m}\sum_{j=1}^m h_j^2 \times 100$$


$$\text{Average System Uniqueness \%} = \frac{1}{m}\sum_{j=1}^m \varphi_j^2 \times 100$$

<br>

**Answers**



In [16]:
# 1. Install the required package
!pip install factor_analyzer -q

import numpy as np
import pandas as pd
from factor_analyzer import FactorAnalyzer

# ==========================================
# Phase 1: Algorithm Definitions
# ==========================================

def decompose_empirical_data(df: pd.DataFrame, k: int) -> dict:
    """Standardizes data, handles guardrails, and extracts latent factors via ML Varimax FA."""
    numeric_df = df.select_dtypes(include=[np.number])
    n, m = numeric_df.shape

    if n <= m:
        raise ValueError(f"Snapshot count (n={n}) must be strictly greater than channel count (m={m}).")
    if k >= m:
        raise ValueError(f"Requested latent space dimension (k={k}) must be strictly less than feature space (m={m}).")

    mu = numeric_df.mean(axis=0)
    sigma = numeric_df.std(axis=0, ddof=1).replace(0.0, 1e-15)
    Z = (numeric_df - mu) / sigma

    fa = FactorAnalyzer(n_factors=k, rotation='varimax', method='ml')
    fa.fit(Z)

    return {
        "Z_matrix": Z.values, # Convert to numpy array for strictly linear algebra later
        "communalities": fa.get_communalities(),
        "uniquenesses": fa.get_uniquenesses(),
        "loadings_matrix": fa.loadings_
    }

def project_latent_scores(Z: np.ndarray, loadings: np.ndarray, h2: np.ndarray, phi2: np.ndarray) -> np.ndarray:
    """Uses Thomson's Regression Method to reconstruct latent states and output telemetry."""
    # 1. State Reconstruction
    R = np.corrcoef(Z, rowvar=False)
    R_inv = np.linalg.inv(R)

    # Thomson's Regression: F = Z * R^-1 * Lambda
    F = Z @ R_inv @ loadings

    # 2. Performance Diagnostic Telemetry
    F_variance = np.var(F, axis=0, ddof=1)

    avg_communality = np.mean(h2) * 100
    avg_uniqueness = np.mean(phi2) * 100

    # Print Telemetry
    print("\n" + "="*50)
    print(" PERFORMANCE DIAGNOSTIC TELEMETRY (THOMSON'S MMSE) ")
    print("="*50)

    print("\n[Reconstructed Factor Scores Empirical Variance]")
    for i, var in enumerate(F_variance):
        print(f"  > Latent Factor {i+1} Variance: {var:.4f}")

    print("\n[System-Wide Allocation Breakdown]")
    print(f"  > Average System Communality : {avg_communality:.2f}%")
    print(f"  > Average System Uniqueness  : {avg_uniqueness:.2f}%")
    print("-" * 50)
    print(f"  > Total Information Conserved: {(avg_communality + avg_uniqueness):.2f}%")
    print("="*50 + "\n")

    return F

# ==========================================
# Phase 2: Execution & Verification
# ==========================================

print("Generating synthetic asset data...")
np.random.seed(42)
n_samples = 3000

# True underlying physics (Latent Factors)
f1 = np.random.normal(0, 1, n_samples)
f2 = np.random.normal(0, 1, n_samples)

# Four sensors observing the structure
s1 = 0.9 * f1 + np.random.normal(0, 0.1, n_samples)
s2 = 0.8 * f1 + 0.2 * f2 + np.random.normal(0, 0.1, n_samples)
s3 = 0.3 * f1 + 0.8 * f2 + np.random.normal(0, 0.1, n_samples)
s4 = 0.1 * f1 + 0.9 * f2 + np.random.normal(0, 1.5, n_samples) # Heavy noise sensor

df_asset = pd.DataFrame({'Sensor_1': s1, 'Sensor_2': s2, 'Sensor_3': s3, 'Sensor_4': s4})

# Run the full pipeline!
print("Extracting subspace (k=2)...")
decomposition_results = decompose_empirical_data(df_asset, k=2)

print("Projecting latent states...")
F_states = project_latent_scores(
    Z=decomposition_results['Z_matrix'],
    loadings=decomposition_results['loadings_matrix'],
    h2=decomposition_results['communalities'],
    phi2=decomposition_results['uniquenesses']
)

print(f"Success! Final projected Factor Matrix Shape: {F_states.shape}")

Generating synthetic asset data...
Extracting subspace (k=2)...
Projecting latent states...

 PERFORMANCE DIAGNOSTIC TELEMETRY (THOMSON'S MMSE) 

[Reconstructed Factor Scores Empirical Variance]
  > Latent Factor 1 Variance: 0.9914
  > Latent Factor 2 Variance: 0.9932

[System-Wide Allocation Breakdown]
  > Average System Communality : 81.03%
  > Average System Uniqueness  : 18.97%
--------------------------------------------------
  > Total Information Conserved: 100.00%

Success! Final projected Factor Matrix Shape: (3000, 2)


/usr/local/lib/python3.12/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(


### Layout Engineering & Interface Design

Render a unified, production-grade **2 × 2 diagnostic canvas** using Plotly Subplots that enforces the following strict geometric layout constraints to ensure maximum readability and prevent overlapping label blocks:

* **Grid Specifications:** Configure the dashboard matrix with an explicit `horizontal_spacing=0.24` and a `vertical_spacing=0.28`. Set the global canvas parameters to `width=1250` and `height=750` utilizing a clean `plotly_white` theme.
* **Legend Formatting:** Place a single, horizontally oriented legend centered directly above the subplots (`orientation="h", x=0.5, y=1.02`). Enforce clean `margin` padding structures to isolate the interface elements (`t=150, b=60, l=140, r=80`).

Populate the grid layout using the four specific tracking profiles below:

```
                  2 × 2 FACTOR DIAGNOSTICS GRID MATRIX
+------------------------------------------+------------------------------------------+
|             [Row 1, Col 1]               |             [Row 1, Col 2]               |
|      Structural Loadings Heatmap         |     Horizontal Stacked Bar Chart         |
|   - Absolute values |λ_(j,r)|            |   - Shared Communality (h²) in Blue      |
|   - Color scale: 'YlOrRd'                |   - Isolated Uniqueness (φ²) in Orange   |
|   - Left-anchored scale bar (x = -0.15)  |   - Total space constraints: 0 to 100%   |
+------------------------------------------+------------------------------------------+
|             [Row 2, Col 1]               |             [Row 2, Col 2]               |
|         Sensor Noise Floor Line          |       Latent Empirical Variance          |
|   - Dot-dashed red line graph with markers|   - Vertical distribution bar chart      |
|   - Tracks exact Uniqueness values (φ²)  |   - Verifies energy level allocations    |
|   - Enforce tick angle adjustments (25°) |   - Black boundary outlines (width = 0.5)|
+------------------------------------------+------------------------------------------+

```

1. **Subplot (1,1): Structural Loadings Matrix Heatmap**
Render the absolute value of the rotated component weights matrix ($|\Lambda_{\text{rotated}}|$) using a continuous `YlOrRd` (Yellow-Orange-Red) color palette. To protect the clarity of the sensor text identifiers along the y-axis, isolate the continuous scale colorbar to the left side of the canvas by explicitly setting `x=-0.15`, `len=0.38`, `y=0.78`, `yanchor="middle"`, and `xanchor="right"`.
2. **Subplot (1,2): Variance Partitioning Profile**
Build a horizontal stacked bar chart mapping out the tracking allocation per sensor channel. The base layer must represent the percentage contribution of shared communalities ($h^2 \times 100$) utilizing color hex code `#1f77b4`. Stack the corresponding sensor uniqueness values ($\varphi^2 \times 100$) directly on top utilizing color hex code `#ff7f0e`. Force the x-axis limits strictly between `[0, 100]`.
3. **Subplot (2,1): Sensor Uniqueness Line Profile**
Construct a clear line chart tracking the isolated channel noise thresholds ($\varphi^2$) across all monitored hardware positions. Apply a red (`#d62728`), 2-pixel wide dot-dashed line string combined with visual `x` indicators. Apply a custom horizontal label layout optimization by rotating the category ticks to a distinct `tickangle=25` constraint.
4. **Subplot (2,2): Latent Factor Variance Distribution**
Draw a vertical bar chart displaying the empirical variance computed across the extracted common factor score vectors. Color the bars green (`#2ca02c`) with an explicit black border envelope (`width=0.5`).

In [17]:
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import numpy as np

# Extract variables from previous steps
loadings_abs = np.abs(decomposition_results['loadings_matrix'])
h2_pct = decomposition_results['communalities'] * 100
phi2_pct = decomposition_results['uniquenesses'] * 100
phi2_raw = decomposition_results['uniquenesses']
f_variance = np.var(F_states, axis=0, ddof=1)

# Labels
sensors = ['Sensor_1', 'Sensor_2', 'Sensor_3', 'Sensor_4']
factors = [f'Factor_{i+1}' for i in range(loadings_abs.shape[1])]

# ==========================================
# Initialize 2x2 Grid with Strict Spacing
# ==========================================
fig = make_subplots(
    rows=2, cols=2,
    horizontal_spacing=0.24,
    vertical_spacing=0.28,
    subplot_titles=(
        "Structural Loadings Matrix |Λ|",
        "Variance Partitioning Profile",
        "Sensor Uniqueness Noise Floor (φ²)",
        "Latent Factor Empirical Variance"
    )
)

# ==========================================
# Subplot (1,1): Structural Loadings Heatmap
# ==========================================
fig.add_trace(
    go.Heatmap(
        z=loadings_abs,
        x=factors,
        y=sensors,
        colorscale='YlOrRd',
        colorbar=dict(
            x=-0.15,            # Left-anchored
            len=0.38,
            y=0.78,
            yanchor="middle",
            xanchor="right",
            title="|Λ|"
        ),
        name="Absolute Loadings"
    ),
    row=1, col=1
)

# Invert y-axis to match standard matrix representation
fig.update_yaxes(autorange="reversed", row=1, col=1)

# ==========================================
# Subplot (1,2): Variance Partitioning Profile
# ==========================================
# Base Layer: Shared Communality (Blue)
fig.add_trace(
    go.Bar(
        y=sensors,
        x=h2_pct,
        name="Shared Communality (h²)",
        orientation='h',
        marker_color='#1f77b4'
    ),
    row=1, col=2
)
# Top Layer: Isolated Uniqueness (Orange)
fig.add_trace(
    go.Bar(
        y=sensors,
        x=phi2_pct,
        name="Isolated Uniqueness (φ²)",
        orientation='h',
        marker_color='#ff7f0e'
    ),
    row=1, col=2
)

# Force x-axis limits and invert y-axis
fig.update_xaxes(range=[0, 100], title_text="Total Variance (%)", row=1, col=2)
fig.update_yaxes(autorange="reversed", row=1, col=2)

# ==========================================
# Subplot (2,1): Sensor Uniqueness Line Profile
# ==========================================
fig.add_trace(
    go.Scatter(
        x=sensors,
        y=phi2_raw,
        mode='lines+markers',
        name="Noise Threshold (φ²)",
        line=dict(color='#d62728', width=2, dash='dashdot'),
        marker=dict(symbol='x', size=8, color='#d62728')
    ),
    row=2, col=1
)

# Enforce explicit tick angle layout constraint
fig.update_xaxes(tickangle=25, row=2, col=1)
fig.update_yaxes(title_text="Uniqueness (φ²)", row=2, col=1)

# ==========================================
# Subplot (2,2): Latent Factor Variance
# ==========================================
fig.add_trace(
    go.Bar(
        x=factors,
        y=f_variance,
        name="Factor Variance",
        marker=dict(
            color='#2ca02c',
            line=dict(color='black', width=0.5) # Black boundary
        )
    ),
    row=2, col=2
)
fig.update_yaxes(title_text="Empirical Variance", row=2, col=2)

# ==========================================
# Global Layout & Architectural Constraints
# ==========================================
fig.update_layout(
    title=dict(
        text="<b>FACTOR DIAGNOSTICS GRID MATRIX</b>",
        font=dict(size=20),
        x=0.5,
        xanchor='center',
        y=0.96
    ),
    width=1250,
    height=750,
    template="plotly_white",
    barmode='stack', # Crucial for Plot 1,2
    legend=dict(
        orientation="h",
        x=0.5,
        y=1.02,
        xanchor="center",
        yanchor="bottom"
    ),
    margin=dict(
        t=150,
        b=60,
        l=140,
        r=80
    )
)

# Display the dashboard inside Colab
fig.show()

# **Subspace Diagnostics & Feature De-correlation**

## **Part A: Develop a Feature Analysis Dashboard**

**UI Layout 1: PCA Optimization Suite ($2\times  3$ Horizontal Layout)**

The first interface must display a 5-panel horizontal grid layout.
* **Panel 1,1 (Heatmap):** Feature Loadings Matrix $|P_{j,r}|$.
* **Panel 1,2 (Bar Chart):** Absolute Eigenvalues ($\lambda_j$) tracking variance magnitude.
* **Panel 1,3 (Combined Bar + Line):** Marginal Explained Variance % per axis (bars) overlaid with Cumulative Captured Variance Curve (dashed line).
* **Panel 2,1 (Bar Chart):** Residual Space (Unexplained Variance %) showing excluded information ($100 - \text{Cumulative Variance \%}$).
* **Panel 2,2 (Scatter Line):** Sample Mean Hotelling's $T^2$ tracking distance vs. Subspace Size $k$.
* **Panel 2,3 (Scatter Line):** Sample Mean $Q$ Statistic (SPE) tracking residual decay vs. Truncation Cutoff $k$.

**UI Layout 2: FA Latent Subspace Suite ($2 \times 2$ Grid Layout)**

The second interface must structure its subplots into a balanced $2 \times 2$ matrix.

* **Panel 1,1 (Heatmap):** Structural Loadings Matrix $|\lambda_{j,r}|$.
* **Panel 1,2 (Stacked Horizontal Bar):** Variance Allocation breakdown showing Shared Communality ($h^2$) versus Channel Uniqueness ($\varphi^2$) totaling $100\%$ per sensor.
* **Panel 2,1 (Scatter Line):** Sensor Uniqueness Noise Floor Profile ($\varphi^2$) across monitored channels to instantly isolate instrument faults.
* **Panel 2,2 (Bar Chart):** Latent Factor Scores Empirical Variance showing the post-rotation power of the extracted physical dimensions.


In [18]:
import numpy as np
import pandas as pd
import plotly.graph_objects as go
from plotly.subplots import make_subplots

# ==========================================
# 1. Generate Data & Extract PCA Subspace
# ==========================================
np.random.seed(42)
n_samples = 3000

# Synthetic Data: 2 underlying forces, 4 sensors (Sensor 4 is noisy)
f1, f2 = np.random.normal(0, 1, n_samples), np.random.normal(0, 1, n_samples)
Z_raw = np.column_stack([
    0.9 * f1 + np.random.normal(0, 0.1, n_samples),
    0.8 * f1 + 0.2 * f2 + np.random.normal(0, 0.1, n_samples),
    0.3 * f1 + 0.8 * f2 + np.random.normal(0, 0.1, n_samples),
    0.1 * f1 + 0.9 * f2 + np.random.normal(0, 1.5, n_samples)
])

# Standardize
Z = (Z_raw - np.mean(Z_raw, axis=0)) / np.std(Z_raw, axis=0, ddof=1)

# Covariance & Eigen Decomposition
S = np.cov(Z, rowvar=False)
eigenvalues, eigenvectors = np.linalg.eigh(S)

# Sort strictly descending
idx = np.argsort(eigenvalues)[::-1]
eigenvalues = eigenvalues[idx]
eigenvectors = eigenvectors[:, idx]

# PCA Projections
Z_pca = Z @ eigenvectors
loadings_abs = np.abs(eigenvectors)

# Variance Calculations
exp_var = (eigenvalues / np.sum(eigenvalues)) * 100
cum_var = np.cumsum(exp_var)
residual_var = 100 - cum_var

# Diagnostics (T2 and Q) across all k cutoffs
k_values = [1, 2, 3, 4]
mean_T2, mean_Q = [], []

for k in k_values:
    if k < 4:
        T2 = np.sum((Z_pca[:, :k]**2) / eigenvalues[:k], axis=1)
        Q = np.sum(Z_pca[:, k:]**2, axis=1)
        mean_T2.append(np.mean(T2))
        mean_Q.append(np.mean(Q))
    else: # If k=4, all variance is explained, Q is technically 0
        T2 = np.sum((Z_pca[:, :k]**2) / eigenvalues[:k], axis=1)
        mean_T2.append(np.mean(T2))
        mean_Q.append(0.0)

# ==========================================
# 2. Build the 2x3 Plotly Dashboard
# ==========================================
components = [f'PC_{i}' for i in k_values]
sensors = [f'S_{i}' for i in k_values]

fig = make_subplots(
    rows=2, cols=3,
    horizontal_spacing=0.12,
    vertical_spacing=0.25,
    subplot_titles=(
        "Feature Loadings Matrix |P|",
        "Absolute Eigenvalues (λ)",
        "Variance Explanation %",
        "Residual Space %",
        "Mean Hotelling's T² vs Subspace (k)",
        "Mean Q Statistic vs Subspace (k)"
    )
)

# [1,1] Heatmap: Absolute Loadings
fig.add_trace(go.Heatmap(
    z=loadings_abs, x=components, y=sensors, colorscale='YlOrRd',
    colorbar=dict(x=-0.1, len=0.42, y=0.78, title="|P|"), name="Loadings"
), row=1, col=1)
fig.update_yaxes(autorange="reversed", row=1, col=1)

# [1,2] Bar Chart: Absolute Eigenvalues
fig.add_trace(go.Bar(
    x=components, y=eigenvalues, marker_color='#9467bd', name="Eigenvalues (λ)"
), row=1, col=2)

# [1,3] Combined: Marginal & Cumulative Variance
fig.add_trace(go.Bar(
    x=components, y=exp_var, marker_color='#1f77b4', name="Marginal %"
), row=1, col=3)
fig.add_trace(go.Scatter(
    x=components, y=cum_var, mode='lines+markers', line=dict(color='#d62728', dash='dash'), name="Cumulative %"
), row=1, col=3)

# [2,1] Bar Chart: Residual Space (Unexplained)
fig.add_trace(go.Bar(
    x=components, y=residual_var, marker_color='#7f7f7f', name="Residual %"
), row=2, col=1)

# [2,2] Scatter Line: Mean Hotelling's T²
fig.add_trace(go.Scatter(
    x=components, y=mean_T2, mode='lines+markers', line=dict(color='#ff7f0e', width=3), name="Mean T²"
), row=2, col=2)

# [2,3] Scatter Line: Mean Q Statistic
fig.add_trace(go.Scatter(
    x=components, y=mean_Q, mode='lines+markers', line=dict(color='#2ca02c', width=3), name="Mean Q"
), row=2, col=3)

# ==========================================
# 3. Global Layout Restrictions
# ==========================================
fig.update_layout(
    title=dict(text="<b>PCA OPTIMIZATION & DIAGNOSTICS SUITE</b>", font=dict(size=22), x=0.5, xanchor='center', y=0.97),
    width=1400, height=800,
    template="plotly_white",
    showlegend=False, # Legend is disabled here to prevent crowding in a 6-panel grid
    margin=dict(t=120, b=60, l=80, r=40)
)

# Apply y-axis limits to percentage charts
fig.update_yaxes(range=[0, 110], row=1, col=3)
fig.update_yaxes(range=[0, 110], row=2, col=1)

fig.show()

In [20]:
# 1. Install necessary library
!pip install factor_analyzer -q

import numpy as np
import pandas as pd
import plotly.graph_objects as go
from plotly.subplots import make_subplots
from factor_analyzer import FactorAnalyzer

# ==========================================
# 2. Generate Data & Extract FA Subspace
# ==========================================
np.random.seed(42)
n_samples = 3000

# Synthetic Data: 2 underlying forces, 4 sensors (Sensor 4 is noisy)
f1 = np.random.normal(0, 1, n_samples)
f2 = np.random.normal(0, 1, n_samples)

s1 = 0.9 * f1 + np.random.normal(0, 0.1, n_samples)
s2 = 0.8 * f1 + 0.2 * f2 + np.random.normal(0, 0.1, n_samples)
s3 = 0.3 * f1 + 0.8 * f2 + np.random.normal(0, 0.1, n_samples)
s4 = 0.1 * f1 + 0.9 * f2 + np.random.normal(0, 1.5, n_samples)

df_asset = pd.DataFrame({'Sensor_1': s1, 'Sensor_2': s2, 'Sensor_3': s3, 'Sensor_4': s4})

# Standardize
mu = df_asset.mean(axis=0)
sigma = df_asset.std(axis=0, ddof=1).replace(0.0, 1e-15)
Z = (df_asset - mu) / sigma

# Fit Factor Analysis
k = 2
fa = FactorAnalyzer(n_factors=k, rotation='varimax', method='ml')
fa.fit(Z)

# Reconstruct Latent States (F) via Thomson's Method
R = np.corrcoef(Z, rowvar=False)
R_inv = np.linalg.inv(R)
F_states = Z.values @ R_inv @ fa.loadings_

# Calculate Metrics for the Dashboard
loadings_abs = np.abs(fa.loadings_)
h2_pct = fa.get_communalities() * 100
phi2_pct = fa.get_uniquenesses() * 100
phi2_raw = fa.get_uniquenesses()
f_variance = np.var(F_states, axis=0, ddof=1)

# ==========================================
# 3. Build the 2x2 Plotly Dashboard (UI Layout 2)
# ==========================================
sensors = ['Sensor_1', 'Sensor_2', 'Sensor_3', 'Sensor_4']
factors = [f'Factor_{i+1}' for i in range(k)]

fig = make_subplots(
    rows=2, cols=2,
    horizontal_spacing=0.24,
    vertical_spacing=0.28,
    subplot_titles=(
        "Structural Loadings Matrix |Λ|",
        "Variance Partitioning Profile",
        "Sensor Uniqueness Noise Floor (φ²)",
        "Latent Factor Empirical Variance"
    )
)

# [1,1] Heatmap: Structural Loadings
fig.add_trace(go.Heatmap(
    z=loadings_abs, x=factors, y=sensors, colorscale='YlOrRd',
    colorbar=dict(x=-0.15, len=0.38, y=0.78, yanchor="middle", xanchor="right", title="|Λ|"),
    name="Absolute Loadings"
), row=1, col=1)
fig.update_yaxes(autorange="reversed", row=1, col=1)

# [1,2] Stacked Bar: Communality vs Uniqueness
fig.add_trace(go.Bar(
    y=sensors, x=h2_pct, name="Shared Communality (h²)", orientation='h', marker_color='#1f77b4'
), row=1, col=2)
fig.add_trace(go.Bar(
    y=sensors, x=phi2_pct, name="Isolated Uniqueness (φ²)", orientation='h', marker_color='#ff7f0e'
), row=1, col=2)
fig.update_xaxes(range=[0, 100], title_text="Total Variance (%)", row=1, col=2)
fig.update_yaxes(autorange="reversed", row=1, col=2)

# [2,1] Scatter Line: Uniqueness Noise Floor
fig.add_trace(go.Scatter(
    x=sensors, y=phi2_raw, mode='lines+markers', name="Noise Threshold (φ²)",
    line=dict(color='#d62728', width=2, dash='dashdot'), marker=dict(symbol='x', size=8, color='#d62728')
), row=2, col=1)
fig.update_xaxes(tickangle=25, row=2, col=1)
fig.update_yaxes(title_text="Uniqueness (φ²)", row=2, col=1)

# [2,2] Bar Chart: Latent Empirical Variance
fig.add_trace(go.Bar(
    x=factors, y=f_variance, name="Factor Variance",
    marker=dict(color='#2ca02c', line=dict(color='black', width=0.5))
), row=2, col=2)
fig.update_yaxes(title_text="Empirical Variance", row=2, col=2)

# ==========================================
# 4. Global Layout Restrictions
# ==========================================
fig.update_layout(
    title=dict(text="<b>FA LATENT SUBSPACE SUITE</b>", font=dict(size=20), x=0.5, xanchor='center', y=0.96),
    width=1250, height=750, template="plotly_white", barmode='stack',
    legend=dict(orientation="h", x=0.5, y=1.02, xanchor="center", yanchor="bottom"),
    margin=dict(t=150, b=60, l=140, r=80)
)

fig.show()

/usr/local/lib/python3.12/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning:

'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.



## **Part B: Analysis**

### Generate the following synthetic data and answer the questions below

The synthetic multi-sensor asset data must be generated using
```python
import numpy as np
import pandas as pd

# Set execution seed for perfect reproducible properties
np.random.seed(42)
n_samples = 2500

# 1. Generate underlying, independent true physical driving forces (Latent Factors)
f1 = np.random.normal(0, 1, n_samples)  # Primary structural mode/load
f2 = np.random.normal(0, 1, n_samples)  # Secondary operational mode

# 2. Construct the Observable Multi-Sensor Array with targeted noise profiles
# Sensors 1 and 2 couple strongly to the primary process (f1)
s1 = 0.85 * f1 + 0.10 * f2 + np.random.normal(0, 0.3, n_samples)
s2 = 0.80 * f1 + 0.15 * f2 + np.random.normal(0, 0.35, n_samples)

# Sensor 3 couples strongly to the secondary process (f2)
s3 = 0.12 * f1 + 0.90 * f2 + np.random.normal(0, 0.25, n_samples)

# Sensor 4 suffers an extreme electrical/instrumentation fault (massive localized white noise)
s4 = 0.02 * f1 + 0.05 * f2 + np.random.normal(0, 1.40, n_samples)

# 3. Compile into the final evaluation DataFrame
df_asset = pd.DataFrame(
    data=np.vstack([s1, s2, s3, s4]).T,
    columns=['Sensor_1', 'Sensor_2', 'Sensor_3', 'Sensor_4']
)

```

**Empirical Covariance and Correlation Properties**

* **Sensors 1, 2, and 3:** High cross-correlation due to strong coupling with the common underlying physical structures ($f_1$ and $f_2$).
* **Sensor 4:** Highly un-correlated with all other channels, but exhibits an exceptionally large variance magnitude ($\sigma^2 \approx 2.0$) due to its high noise contamination scale.


### Question 1: The Total Variance Illusion (PCA Subplots 1 & 2 vs. FA Subplots 1 & 2)

* **Context:** In your PCA dashboard, `PC 1` captures roughly $45\%$ of the total system variance, and `PC 2` captures over $30\%$, indicating a highly structured, multi-component system. However, looking at the Factor Analysis dashboard, the **Variance Partitioning Stacked Bar** reveals that `Sensor_4` has a **Uniqueness value ($\varphi^2$) exceeding $98\%$**.
* **Core Problem:** 1. What does a Uniqueness value close to $100\%$ tell you about the physical nature of the data coming from `Sensor_4`?
2. Because PCA maximizes total global variance without distinguishing between types of variance, how does the massive localized noise of `Sensor_4` "trick" the PCA engine into inflating its top eigenvalues?
3. Why is this a major risk if an engineer relies *only* on PCA to build an automated plant anomaly detection framework?



<br>

**Answers**

1. A Uniqueness value ($\varphi^2$) close to 100% indicates that nearly all the variance produced by Sensor_4 is completely isolated and idiosyncratic. It shares no meaningful underlying correlation with the rest of the physical structure. In the real world, this represents pure instrument noise, an electrical short, or a severed connection, rather than an actual structural load.

2. PCA is a geometric projection that maximizes total global variance, without caring where that variance comes from. Because Sensor_4 has a massive localized variance ($\sigma^2 \approx 2.0$), PCA interprets this loudness as mathematical importance. It will dedicate a top principal component specifically to align with this massive noise vector, artificially inflating its top eigenvalues ($\lambda$) and confusing pure noise with a major structural signal.

3. If we relie solely on PCA, the automated anomaly detection framework will map this localized sensor fault directly into the primary healthy subspace. The system will become highly sensitive to electrical noise from a single bad sensor, triggering massive false-positive plant alarms while potentially missing subtle but dangerous degradation in the true structural components.

<br>


### Question 2: Decoupling Structural Loading via Rotation (FA Subplot 1 vs. PCA Eigenvectors)

* **Context:** Look at the **Structural Loadings Matrix** heatmap in the top-left of the FA dashboard. You can see that `Sensor_1` and `Sensor_2` load heavily on `Factor 1`, while `Sensor_3` loads exclusively on `Factor 2`.
* **Core Problem:**
1. Traditional PCA forces its principal axes to follow a strict mathematical hierarchy ($\lambda_1 > \lambda_2 > \dots$) where every component tries to sweep up as much mixed variance as possible across all sensors. Explain how the **Varimax rotation** in Factor Analysis relaxes this rule to create "simple structure" (clean, isolated groupings).
2. From a plant operator's standpoint, why is the rotated FA heatmap much easier to troubleshoot during a structural failure than a raw, unrotated PCA eigenvector matrix?





<br>

**Answers**

1. Traditional PCA forces a strict mathematical hierarchy where the first component sweeps up the maximum possible variance across all sensors, resulting in messy, mixed loadings where every sensor contributes a bit to every component. Varimax rotation in Factor Analysis relaxes this rigid rule. It rotates the axes to maximize the variance of the squared loadings themselves, forcing coefficients toward 1 or 0. This creates a simple structure where each sensor aligns cleanly and exclusively with a single specific physical factor.

2. From a plant operator's perspective, simple structure maps directly to physical reality. Because Varimax rotation isolates Sensor_1 and Sensor_2 to Factor 1, and Sensor_3 to Factor 2, operators can track specific subsystems. If a structural failure causes the real-time score for Factor 2 to spike, the maintenance team knows exactly to look the physical zone that monitored by Sensor_3. Raw PCA eigenvectors mix the signals, making it nearly impossible to trace a statistical anomaly back to a specific physical beam or bearing.

<br>


### Question 3: Determining Subspace Truncation ($k$) using $T^2$ and $Q$ Profiles

* **Context:** Look at the right side of the PCA dashboard, showing **Mean Hotelling's $T^2$** and **Mean $Q$ Statistic (SPE)** plotted against the truncation cutoff $k$.
* **Core Problem:**
1. Describe the exact behavior of the Mean $Q$ Statistic curve as the subspace cutoff transitions from $k=1$ to $k=2$, and then from $k=2$ to $k=3$.
2. Why does a sharp drop followed by a distinct "flat elbow" in the $Q$ statistic identify **$k=2$** as the true hidden physical dimensionality of this asset?
3. If you incorrectly choose a truncation cutoff of $k=3$ for your monitoring system, what type of data are you forcing into your "clean" principal subspace?

<br>

**Answers**

1. As the subspace transitions from $k = 1$ to $k = 2$, the Mean $Q$ statistic will experience a sharp, dramatic drop. This occurs because the model has just incorporated the second major physical driving force $f_2$. As the cutoff moves from $k = 2$ to $k = 3$, the drop in the $Q$ statistic becomes very marginal and completely flattens out.

2. The sharp drop followed by the distinct flat elbow identifies $k = 2$ as the true hidden physical dimensionality of the asset. The drop confirms the capture of deterministic structural signals, and the flat line confirms that any subsequent dimensions are completely devoid of shared structural information.

3. If you incorrectly choose a truncation cutoff of $k = 3$, you force unstructured, localized random noise into your clean principal subspace. This dilutes the statistical power of your Hotelling's $T^2$ metric. Your baseline definition of healthy now officially includes a high degree of random instability, making the system blind to real structural shifts.

<br>


### Question 4: Operational Trade-offs in System Health Monitoring

* **Context:** Imagine you are deploying a real-time predictive maintenance pipeline for a fleet of identical industrial assets based on these exact models.
* **Core Problem:** Compare the operational trade-offs of these two approaches:
1. **The PCA Strategy:** Tracking the main principal subspace using Hotelling's $T^2$ alongside a residual noise floor monitor using the $Q$ statistic.
2. **The FA Strategy:** Monitoring the rotated Latent Factor Scores directly.


* Which strategy is more robust against a single sensor losing calibration or experiencing an electrical short? Justify your choice using the **Sensor Uniqueness Noise Floor ($\varphi^2$)** metrics.

<br>

**Answers**

1. The PCA Strategy is computationally simpler and faster, defining health purely by geometric distance,$T^2$ for the main subspace, $Q$ for the residual. However, as demonstrated by Sensor_4, it is highly vulnerable to being skewed by loud, localized noise masquerading as structural variance.

2. The FA Strategy requires more parametric setup and mathematical assumptions, but it defines health by explicitly separating shared system behavior Latent Scores from isolated sensor noise as Uniqueness.

The FA Strategy is significantly more robust against a single sensor losing calibration or experiencing an electrical short.

Because Factor Analysis explicitly accounts for idiosyncratic variance through the Sensor Uniqueness ($\varphi^2$) metric, it effectively filters out localized faults before they can corrupt the main monitoring layer. If Sensor_4 experiences a massive electrical short, FA absorbs that variance directly into $\varphi_4^2$. The shared Latent Factor scores remain perfectly stable and accurate, preventing a false plant shutdown. Operators can simply set an alarm on the $\varphi^2$ metrics to instantly flag and replace broken sensors, while the structural monitoring continues uninterrupted.